orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -869,89 +869,217 @@ common::Status HandleSharedInitializer(Graph& graph,
+                                            true, /* initializer data */
+                                            send_input_args,
+                                            new_input_names));
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<size_t>(graph,
+                                              "send_dst_rank" + cut_index_str,
+                                              ONNX_NAMESPACE::TensorProto_DataType_INT64,
+                                              next_stage, /* initializer data */
+                                              send_input_args,
+                                              new_input_names));
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<bool>(graph,
+                                            "recv_input_signal" + cut_index_str,
+                                            ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                                            true, /* initializer data */
+                                            recv_input_args,
+                                            new_input_names));
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<size_t>(graph,
+                                              "recv_src_rank" + cut_index_str,
+                                              ONNX_NAMESPACE::TensorProto_DataType_INT64,
+                                              current_stage, /* initializer data */
+                                              recv_input_args,
+                                              new_input_names));
+
+  // add output node_arg for send/recv
+  AddNewNodeArg(graph, "send_output_signal" + cut_index_str,
+                ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                send_output_args, new_output_names);
+
+  AddNewNodeArg(graph, "receive_output_signal" + cut_index_str,
+                ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                recv_output_args, new_output_names);
+
+  return Status::OK();
+}
+
+common::Status SplitGraphWithMap(Graph& graph,
+                                 std::map<Node*, int>& op_to_stage,
+                                 int nstages,
+                                 std::vector<std::pair<int, int>>& messages,
+                                 std::vector<Node*>& send_nodes,
+                                 std::vector<Node*>& recv_nodes) {
+
+  // forward_messages[s]: all the tensors sent by stage s while executing
+  // forward computation.
+  std::vector<std::set<const NodeArg*>> forward_messages(nstages);
+  // TODO(jufranc): once we start using this function on the training graph,
+  // we need to keep backward_messages[s] too.
+
+  // Tensors that need to be sent from one device to the other.
+  // TODO(jufranc): Should we consider weights here too? 
+  // forwarded_tensors[i] = {t, {stage of producer, stage of the last consumer}}
+  std::vector<std::pair<const NodeArg*, std::pair<int, int>>> forwarded_tensors;
```

Comment:
// forwarded_tensors[i] = {t, {stage of t's producer, stage of the t's last consumer}}

Reply:
hopefully, the new comments explain it better.

Reply:
Thanks. It looks great. I don't have a specific format but I feel it's sufficient to explain `collection` by specifying the meaning of its individual element (just like what you did) and avoiding undefined symbols and words.

---
In reply to: [537624140](https://github.com/microsoft/onnxruntime/pull/5940#discussion_r537624140) [](ancestors = 537624140)